src/main/scala/com/fulcrumgenomics/util/Metric.scala

```diff
@@ -0,0 +1,129 @@
+    val clazz             = ReflectionUtil.typeTagToClass[T]
+    val reflectiveBuilder = new ReflectiveBuilder(clazz)
+    val numArgsRequired   = reflectiveBuilder.argumentLookup.view.length
+    val numArgsFound      = argValues.length
+    if (numArgsRequired != numArgsFound) {
+      throw new IllegalArgumentException(s"Expected $numArgsRequired arguments but found $numArgsFound.")
+    }
+    val arguments = reflectiveBuilder.argumentLookup.ordered.zip(argValues).map { case (arg, value) =>
+      ReflectionUtil.constructFromString(arg.argumentType, arg.unitType, value) match {
+        case Success(v) => v
+        case Failure(thr) => throw thr
+      }
+    }
+    reflectiveBuilder.build(arguments)
+  }
+
+  /** Reads metrics from a set of lines.  The first line should be the header with the field names.  Each subsequent
+    * line should be a single metric. */
+  def read[T <: Metric](lines: Iterator[String], source: Option[String] = None)(implicit tt: ru.TypeTag[T]): Seq[T] = {
+    if (lines.isEmpty) throw new IllegalArgumentException(s"No header found in metrics" + source.map(" in source: " + _).getOrElse(""))
+    val names = lines.next().split(Delimiter)
+    argNames[T].zip(names).foreach { case (expectedName, actualName) =>
+      if (expectedName != actualName) {
+        throw new IllegalStateException(
+          s"Header out of order (found '$actualName' but expected '$expectedName' in metrics" +
+          source.map(" in source: " + _).getOrElse("")
+        )
+      }
+    }
+    if (lines.isEmpty) Seq.empty
+    else {
+      lines.map { line =>
+        val arguments = line.split(Delimiter)
+        build[T](arguments)
+      }.toSeq
+    }
+  }
+
+  def read[T <: Metric](path: Path)(implicit tt: ru.TypeTag[T]): Seq[T] = read[T](Io.readLines(path), Some(path.toString))
+
+  /** Writes a metric to the given writer.  The first line will be a header with the field names.  Each subsequet
+    * line is a single metric.
+    */
+  def write[T <: Metric](metrics: Seq[T], writer: Writer)(implicit tt: ru.TypeTag[T]): Unit = {
+    writer.write(argNames[T].mkString(Delimiter))
+    writer.write("\n")
+    metrics.foreach { metric =>
+      writer.write(metric.argValues.mkString(Delimiter))
+      writer.write("\n")
+    }
+  }
+
+  def write[T <: Metric](metrics: Seq[T], path: Path)(implicit tt: ru.TypeTag[T]): Unit = {
+    val writer = Io.toWriter(path)
+    write(metrics, writer)
+    writer.close()
+  }
+}
+
+/**
+  * Base trait for metrics.
+  *
+  * All classes extending this class should be a case class.
+  */
+trait Metric {
+  private lazy val reflectiveBuilder = new ReflectiveBuilder(this.getClass)
+
+  /** Get the names of the arguments in the order they were defined. */
+  def argNames: Seq[String] = {
+    this.reflectiveBuilder.argumentLookup.ordered.map(_.name)
+  }
+
+  /** Get the values of the arguments in the order they were defined. */
+  def argValues: Seq[String] = {
```

Comment:
And here `values`?  Also, I think given the requirement that subclasses be case classes, if there's some way in the type system to require that then this could be implemented as `productIterator.map(_.toString).toSeq`